Sources/ShapedArray/Basic.swift

```diff
@@ -108,16 +110,20 @@ extension ShapedArray {
     @usableFromInline
     internal func isValid<T: BinaryInteger>(axis k: T) -> Bool {
         let axis = Int(k)
-        return axis >= -rank && axis < rank
+        return axis >= -self.rank && axis < self.rank
     }
-    
-    /// Returns `true` iff each element of `axes` denotes an axis of `self`.
+
+    /// Returns `true` iff each element of `axes` denotes an axis of `self`
+    /// _and_ there is no repeated axis.
     @usableFromInline
     internal func areValid<T: BinaryInteger>(axes: [T]) -> Bool {
-        return axes.allSatisfy { isValid(axis: $0) }
+        let rank = self.rank
+        return axes.allSatisfy { self.isValid(axis: $0) }
+            && (Set(axes.map { $0 < 0 ? Int($0) + rank : Int($0) }).count == axes.count)
     }
-    
-    /// Returns `true` iff each element of `axes` denotes an axis of `self`.
+
+    /// Returns `true` iff each element of `axes` denotes an axis of `self`
+    /// _and_ there is no repeated axis.
```

Comment:
realized it doesn't make sense to have axes that have a repeated axis, so added that to the precondition. that may be something we wish to assert on so it's optimized away though since it adds a `Set` construction